### PR TITLE
Add periodic variable coefficient linear advection equation

### LINKED ISSUE #176: Batteries included: variable coefficient linear advection in periodic domain
There is already a pre-made setup for nonperiodic domains. It would be nice to mimic this for periodic domains.

### PR DESCRIPTION
**Pull request by JoshuaLampert**
This adds a semidiscretization for the periodic variable coefficient linear advection equation.
Closes #176.

### CODE REVIEW

File: docs/src/introduction.md [hunk 1]
```diff
@@ -336,7 +336,7 @@ equations (PDEs). These are shipped with this package and you are encouraged to
 look at their source code to learn more about it.
 
 - Linear scalar advection with variable coefficient:
-  [`VariableLinearAdvectionNonperiodicSemidiscretization`](@ref)
+  [`VariableLinearAdvectionNonperiodicSemidiscretization`](@ref), [`VariableLinearAdvectionPeriodicSemidiscretization`](@ref)
```
Comment (ranocha) [inline docs/src/introduction.md:339]:
```suggestion
  [`VariableLinearAdvectionPeriodicSemidiscretization`](@ref), [`VariableLinearAdvectionNonperiodicSemidiscretization`](@ref)
```
The periodic case is conceptually simpler, so I would like to present it first.

**Review by ranocha — CHANGES_REQUESTED** [review]
Great, thanks! Did you run it for some non-trivial setup to check whether it works as intended?

**Comment by JoshuaLampert** [comment]
Yes, I performed some tests, e.g. with constant velocity and piecewise constant velocity (mostly because the solution is easy to verify). The solutions seemed to be reasonable. If you have a specific velocity function with corresponding reference solution in mind I'd be happy to test that, as well.

**Comment by ranocha** [comment]
There seems to be a problem with the error-based step size controller in my setup. Using a fixed step size as in
```julia
sol = solve(ode, SSPRK104(), dt=0.8*D.Δx, adaptive=false, save_everystep=true)
```
works better. Then, you have a discontinuous coefficient, so any standard high-order finite difference method will result in oscillations - in particular at the periodic boundaries, where the mass is concentrated (instead of flowing out of the domain as in the nonperiodic case).

There is also a problem at the discontinuity at zero. Are you sure that your solution is correct? It does not seem to conserve the total mass $\int u$.

**Comment by ranocha** [comment]
You can find a discussion of related issues in the paper "Accuracy of Stable, High-order Finite Difference Methods
for Hyperbolic Systems with Non-smooth Wave Speeds" (https://doi.org/10.1007/s10915-019-01088-w)

**Comment by JoshuaLampert** [comment]
Thanks for the clarification! I expected the oscillations, but was wondering about the discontinuity at 0. I naively assumed the analytical solution would be the same for a piecewise constant coefficient as for a constant one, but you're right. It doesn't conserve the total mass.
So this is fine, I guess?

**Comment by ranocha** [comment]
Yes. Let's go ahead and merge this PR. Thanks!

**Pull request merged**